### PR TITLE
Add libxkbregistry to query available RMLVO

### LINKED ISSUE #123: Provide an API for the evdev.xml registry
**Note: this is a discussion topic for now, not an issue**

Right now, tools parse `evdev.xml` to get the list ("registry") of all possible models, layouts and their variants, and options. Those are then displayed to the user who selects a subset of them and that selection is then passed as RMLVO to libxkbcommon.

These days the rules are usually hardcoded as "evdev" for historical reasons. For the same reasons, `evdev.xml` is hardcoded as the path to parse.

Ideally, libxkbcommon should/could provide this registry as well so we don't rely on a third-party source and parser.

A few notes:
- `evdev.xml` is generated by xkeyboard-config by merging the various parts xml together. A second `evdev.extras.xml` is available for those options that are too niche to be included in the normal xml file.
- `evdev.lst` is the deprecated older format and generated from `evdev.xml` but the conversion is lossy. Specifically, it drops `allowMultipleSelection="true"` and a few other lesser-used tags. 
  - There is no `evdev-extras.lst`.

Implementation-wise I see two options:
- a full API that provides the information from the xml file, e.g. `xkb_registry_layout_get_variants()` to get the list of variants from the registry.
- a simpler API that passes a file descriptor to the underlying XML file. That leaves the parsing up to the client and we merely control what they open.

The latter is simpler right now but gets trickier with #108 merged. If we want to allow for `$XDG_CONFIG_HOME/xkb/rules/evdev.xml` (and we should to have this show up in GUIs) we need to merge the XML files together ourselves and then provide the fd to the merged XML file.

Final comment: this is mostly independent of libxkbcommon so it could be a separate library altogether. The primary reason why it should be part of libxkbcommon is that it should share the same include paths so the list always matches what libxkbcommon can actually load. That could be achieved by having a `libxkbcommon-registry.so`. Or a completely external library that takes a `struct xkb_context` and calls `xkb_context_include_path_get()` and then goes from there.

Comments?

### PR DESCRIPTION
**Pull request by whot**
This library is the replacement for clients parsing `evdev.xml` directly.
Instead, they should use the API here so that in the future we may even
be able to swap evdev.xml for a more suitable data format.

A few notes on the implementation here:
- I still need to write the code in gnome-control-center to have a real world user, so far I've only toyed around with examples
- the API is iterator-based, mostly because every client is expected to iterate through the various bits and this makes it easier than passing arrays of pointers around for those clients
- A lot of the actual code is generated by macros. We can eventually swap this for manually-typed code but right now it's easier to iterate over given that all objects are nearly identical
- still needs more testing, this is the first working version copied over from https://github.com/whot/libxkbregistry/issues/1

cc @svu, @fooishbar 

Fixes #123

### CODE REVIEW

**Comment by whot** [comment]
Couldn't figure out what the magic is to get the macos and windows builds to actually compile, so I've disabled it on both for now.

**Comment by whot** [comment]
Ok, marking this as ready for review. There's still work to be done (more tests, better error handling) but I think overall it's ready for comments. There's a (untested) port of [gnome-desktop](https://gitlab.gnome.org/whot/gnome-desktop/-/commits/wip/xkbregistry) to the new API and the [example in this patchset](https://github.com/whot/libxkbcommon/blob/wip/libxkbregistry/examples/list-all.c).

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ jobs:
           HOMEBREW_NO_INSTALL_CLEANUP: 1
       - name: Setup
         run: |
-          PATH="/usr/local/opt/bison/bin:${PATH}" meson setup -Denable-wayland=false -Denable-x11=false build
+          PATH="/usr/local/opt/bison/bin:${PATH}" meson setup -Denable-wayland=false -Denable-x11=false -Denable-xkbregistry=false build
```
[the remark this answers is not in the document]
Comment (whot) [reply .github/workflows/main.yml:56]: tried all that, didn't work. with/without force, changing `LD_LIBRARY_PATH`, etc. Not sure what else I can try atm - I have a mac but it's not set up for development so I couldn't reproduce this locally either. libxml2 is supposed to be already there anyway but neither the native one nor the brew one gets picked up correctly.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -76,7 +76,7 @@ jobs:
         shell: cmd
         run: |
           call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
-          meson setup -Denable-wayland=false -Denable-x11=false -Denable-docs=false build
+          meson setup -Denable-wayland=false -Denable-x11=false -Denable-docs=false -Denable-xkbregistry=false build
```
[the remark this answers is not in the document]
Comment (whot) [reply .github/workflows/main.yml:79]: libxml2 doesn't seem to be in `choco`, so tbh I'm not sure what to do here. never done windows development before.

**Comment by whot** [comment]
thx for the reviews, I'll fix the code asap. recording the general comments:

>  you opted to make all of the objects standalone heap-allocated reference counted object

basically: it was simplest at the time. I expect this one to be a zero-maintenance library so the simplest path was the easiest. You could argue that it's even simpler to have just one object but when I wrote this code first the API looked different ;)

Either way, long term ref/unref locks us less into a specific behaviour than having just one immutable object and I can't think of a use-case where performance and allocation types matter - it's all either generating a list once or presenting the data to the user in which case the user is the slowest bit anyway.

As for `PACKAGING` - completely forgot about that. I think splitting it out is "cleaner", there are only a handful of applications that need the registry but libxkbcommon is quite, ahaha, common. ;)

File: src/registry.c [hunk 3]
```diff
@@ -0,0 +1,914 @@
...
+        snprintf(rules, sizeof(rules), "%s/rules/%s.xml", *path, ruleset);
+        success = parse(ctx, rules, RXKB_POPULARITY_STANDARD);
+
+        if (success) {
+            if (ctx->load_extra_rules_files) {
+                /* Where we find a directory with our wanted ruleset xml files
+                 * and that ruleset parses successfully, parse the .extras.xml
+                 * file as well (doesn't matter if it doesn't succeed).
+                 *
+                 * FIXME: should only do this for the system directories, I
```
[the remark this answers is not in the document]
Comment (whot) [reply src/registry.c:506]: long-term we should get rid of the `extras.xml` file and use the existing but unused `popularity` tag instead (see the [xkeyboard-config merge request](https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/merge_requests/76)). We might never be able to get rid of this behaviour for system directories, but we can force the right thing for the custom includes.

and, tbh, I'm also having a hard time thinking of a use-case where you want to define keyboard layouts outside the system directories but you *don't* want them to show up in the default list.

File: xkbcommon/xkbregistry.h [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (whot) [reply xkbcommon/xkbregistry.h:25]: hehe. you started using it when you switched to meson - it uses `pragma once` in the generated `config.h` file :)

File: xkbcommon/xkbregistry.h [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+ * libxkbcommon.
+ *
+ * This library is the replacement for clients currently parsing evdev.xml
+ * directly.
+ *
+ * @{
+ */
+
+/**
+ * @struct rxkb_context
```
[the remark this answers is not in the document]
Comment (whot) [reply xkbcommon/xkbregistry.h:53]: two reasons - rxkb flows easier (for unknown reasons) but the main selling point is that  it is immediately visually distinct from xkbcommon's `xkb_` prefixes. Compare `xkbr_context_new()` and `rxkb_context_new()`.

File: xkbregistry.map [hunk 6]
```diff
@@ -0,0 +1,60 @@
+/* versions are kept in sync with libxkbcommon.so */
```
[the remark this answers is not in the document]
Comment (whot) [reply xkbregistry.map:1]: ack, needs special treatment though since a lot of the functions are macro-generated and don't show up in grep. Need to think about this.

File: xkbcommon/xkbregistry.h [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+ * object hits zero, associated resources will be freed.
+ *
+ * @param ctx The xkb registry context
+ * @return always NULL
+ */
+struct rxkb_context*
+rxkb_context_unref(struct rxkb_context *ctx);
+
+/**
+ * Assign user-specific data. libxkbregistry will not look at or modify the
```
[the remark this answers is not in the document]
Comment (whot) [reply xkbcommon/xkbregistry.h:214]: I honestly don't know. Thought it was easier to add since it's a virtual noop for us but happy to drop it.

File: meson.build [hunk 8]
```diff
@@ -296,6 +296,51 @@ You can disable X11 support with -Denable-x11=false.''')
     )
 endif
 
+# libxkbregistry
+if get_option('enable-xkbregistry')
+    dep_libxml = dependency('libxml-2.0')
```
[the remark this answers is not in the document]
Comment (whot) [reply meson.build:301]: libxml2 has support for `XInclude` directly available, so it's just a matter of defining the callbacks. expat requires [manual handling of the tags](https://github.com/libexpat/libexpat/blob/master/expat/doc/reference.html#L751).

To keep the code simple and the feature flexible, I went with `XInclude`. For example, my current test config file looks like this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE xkbConfigRegistry SYSTEM "xkb.dtd">
<xkbConfigRegistry version="1.1" xmlns:xi="http://www.w3.org/2001/XInclude">
 <layoutList>
   <layout>
     <configItem>
       <name>foo</name>
        <shortDescription>foo</shortDescription>
       <description>A totally foo item</description>
       <languageList>
         <iso639Id>eng</iso639Id>
       </languageList>
     </configItem>
   </layout>
   <xi:include href="xkb://rules/evdev.xml" xpointer="xpointer(xkbConfigRegistry/layoutList/*)" />
 </layoutList>

```

[the remark this answers is not in the document]
Comment (whot) [reply meson.build:301]: sorry, hit enter too early, amended the comment above now

(this comment is just for email benefit, will delete it :)

File: xkbregistry.map (hunk 6 above)
[the remark this answers is not in the document]
Comment (bluetech) [reply xkbregistry.map:1]: TBH I'm not a huge fan of the function-generating macros, because they break grep and jump-to-definition and such. Sure there's a lot of duplication but I'm not sure it really matters. (But I totally defer to your discretion on that).

**Comment by bluetech** [comment]
Regarding the ref/unref, makes sense.

Regarding PACKAGING, I agree, let's recommend to split it.

File: src/registry.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (whot) [reply src/registry.c:506]: `gnome-desktop` uses dconf key `"org.gnome.desktop.input-sources` `show-all-sources` (toggle in `gnome-tweaks`) to decide whether to load it or not. I don't know how often it's used but generally if GNOME decides to expose something, someone's has asked for it :) I think say that's prominent enough and a large enough user base.

Either way, since we're the intermediate here it's much easier for us to just pass that detail on and let the caller decide whether to ignore it or not. And as I said, it won't matter for virtually all use-cases anyway.

File: meson.build (hunk 8 above)
[the remark this answers is not in the document]
Comment (whot) [reply meson.build:301]: I've played with a few things but had other things take priority last week. The main thing I tried was to validate input through the DTD (so we don't have to check things like the name field manually). That was painful (`XInclude` got in the way) but got it to work eventually (will push after cleaning up). The second thing was that `XInclude` is a dead end anyway.

I *suspect* that the two majority use-cases will be adding an option and adding a variant. The latter cannot be handled with pure `XInclude` anyway because you can't  easily merge a new variant *into* an existing system layout. So some other merge strategy becomes inevitable sooner than later and then we just parse two separate XML files. I'm not 100% sure on the exact behaviour and whether it'll be an explicit `<merge>` tag or implied behaviour.

So in summary: `XInclude` will go away, and libxml vs expat can be revisited then too.

File: xkbregistry.map (hunk 6 above)
[the remark this answers is not in the document]
Comment (whot) [reply xkbregistry.map:1]: We can expand the macros to normal functions once we're happy with the code. For now it's a lot easier to change one macro than to update 10 `ref/unref` calls :)

File: xkbcommon/xkbregistry.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (whot) [reply xkbcommon/xkbregistry.h:214]: actually, got a use-case now: user data is needed to pass caller data through to the log function (i copied the log system from libxkbcommon)

**Comment by whot** [comment]
I've pushed the current state in case you manage to find time for a review, but it contains the list of changes (cleaned up a bit) so you can follow the history too. So looking at the full diff is going to be more useful than per-commit (but at least for now the commit history is useful in its own regard, at least for me).

Short outline of the history is: added DTD parsing for the current code, but dropped it again before implementing the new merging behaviour (see ba9745e). DTD validation then made a comeback adjusted for that behaviour, with manual XML checking added and then reverted again.

I think this contains all straightforward changes you requested in the original review too, anything I haven't specifically commented on is fixed.

**Comment by whot** [comment]
All squashed down into one commit now because tbh it was getting to unwieldly and the extra commits have little information beyond some experimental lessons. This has tests now (not 100% coverage) and the new merging code which works so that:
- system files are parsed first
- user files **append** to the system data

You can't overwrite data because I don't think there's a use-case for it and this limitation allows a user to provide a minimal xml file to append e.g. their own variant for an existing system layout (e.g. "us(banana)" or something).

XML validation is done through the DTD because it's easiest that way and the least amount of custom code.
A few error checks were added too so now e.g. `parse()` will fail if you run it twice. The public API has not changed other than that extra error checking.

At this point I'm struggling to see any bugs in this code because I've looked at it for too long. Any review would be appreciated.

**Comment by whot** [comment]
yay, windows now passes too :) albeit you ma not be happy with the fix for that...

File: src/registry.c [hunk 9]
```diff
@@ -0,0 +1,1196 @@
...
+
+    darray_foreach_reverse(path, ctx->includes) {
+        char rules[PATH_MAX];
+
+        if (snprintf_safe(rules, sizeof(rules), "%s/rules/%s.xml",
+                           *path, ruleset)) {
+            log_dbg(ctx, "Parsing %s\n", rules);
+            if (parse(ctx, rules, RXKB_POPULARITY_STANDARD))
+                success = true;
+        }
```
Comment (bluetech) [inline src/registry.c:657]: Maybe log an error if the snprintf fails?

File: src/registry.c [hunk 10]
```diff
@@ -0,0 +1,1196 @@
...
+static void
+ATTR_PRINTF(2, 0)
+xml_error_func(void *ctx, const char *msg, ...)
+{
+    static char buf[PATH_MAX];
+    static int slen = 0;
+    va_list args;
+    int rc;
+
+    /* libxml2 prints IO errors from bad includes paths by
```
Comment (bluetech) [inline src/registry.c:1066]: Is this still relevant now that we don't do includes anymore?

File: src/registry.c [hunk 11]
```diff
@@ -0,0 +1,1196 @@
...
+
+static bool
+validate(struct rxkb_context *ctx, xmlDoc *doc)
+{
+    bool success = false;
+    xmlValidCtxt *dtdvalid = NULL;
+    xmlDtd *dtd = NULL;
+    xmlParserInputBufferPtr buf = NULL;
+    /* This is a modified version of the xkeyboard-config xkb.dtd, we
+     * allow for more either of modelList, layoutList or optionList to be
```
Comment (bluetech) [inline src/registry.c:1106]: I couldn't quite figure out what this says

File: src/registry.c [hunk 12]
```diff
@@ -0,0 +1,1196 @@
...
+     * missing.
+     */
+    const char dtdstr[] =
```
Comment (bluetech) [inline src/registry.c:1109]: My C-fu is a little weak -- will it put this in rodata or on the stack? I'd prefer if it was in rodata, I don't remember if that requires `const char *` + `strlen` or if `const char[]` + `sizeof` also works.

File: src/registry.c [hunk 13]
```diff
@@ -0,0 +1,1196 @@
...
+        "<!ELEMENT description (#PCDATA)>\n"
+        "<!ELEMENT vendor (#PCDATA)>\n"
+        "<!ELEMENT countryList (iso3166Id+)>\n"
+        "<!ELEMENT iso3166Id (#PCDATA)>\n"
+        "<!ELEMENT languageList (iso639Id+)>\n"
+        "<!ELEMENT iso639Id (#PCDATA)>\n"
+        "<!ELEMENT hwList (hwId+)>\n"
+        "<!ELEMENT hwId (#PCDATA)>\n";
+
+    /* Note: do not use xmlParserInputBufferCreateStatic, it generates random
```
Comment (bluetech) [inline src/registry.c:1136]: Hehe

File: src/registry.c [hunk 14]
```diff
@@ -0,0 +1,1196 @@
...
+
+    dtd = xmlIOParseDTD(NULL, buf, XML_CHAR_ENCODING_UTF8);
+    if (!dtd) {
+        log_err(ctx, "Failed to load DTD\n");
+        return false;
+    }
+
+    dtdvalid = xmlNewValidCtxt();
+    if (!xmlValidateDtd(dtdvalid, doc, dtd)) {
+        log_err(ctx, "XML error: failed to validate document\n");
```
Comment (bluetech) [inline src/registry.c:1151]: Can we include the path in this message?

File: src/registry.c [hunk 15]
```diff
@@ -0,0 +1,1196 @@
...
+        "<!ELEMENT iso639Id (#PCDATA)>\n"
+        "<!ELEMENT hwList (hwId+)>\n"
+        "<!ELEMENT hwId (#PCDATA)>\n";
+
+    /* Note: do not use xmlParserInputBufferCreateStatic, it generates random
+     * DTD validity errors for unknown reasons */
+    buf = xmlParserInputBufferCreateMem(dtdstr, sizeof(dtdstr),
+                                        XML_CHAR_ENCODING_UTF8);
+    if (!buf)
+        return false;
```
Comment (bluetech) [inline src/registry.c:1141]: I wonder if for this and the next error we should return `true` instead, because it is some internal error and not actually a validation error, so graceful fallback makes sense. Hmm probably not.

File: src/registry.c [hunk 16]
```diff
@@ -0,0 +1,1196 @@
...
+      enum rxkb_popularity popularity)
+{
+    bool success = false;
+    xmlDoc *doc = NULL;
+    xmlNode *root = NULL;
+
+    if (!check_eaccess(path, R_OK))
+        return false;
+
+    LIBXML_TEST_VERSION
```
Comment (bluetech) [inline src/registry.c:1176]: What does this do?

File: src/registry.c (hunk 9 above)
Comment (whot) [reply src/registry.c:657]: Do we need to worry about a use-case where we exceed `PATH_MAX` for file length? I'm pretty sure all sorts of funny issues happen then. For the default case (system + `XDG_CONFIG_HOME`) we should never hit this and `rxkb_context_include_path_append()` has a check for this case so it won't let you append the path to begin with. There are some cases to get around it but IMO they're not worth worrying about.

File: src/registry.c (hunk 12 above)
Comment (whot) [reply src/registry.c:1109]:
```
objcopy build/libxkbregistry.so /dev/null --dump-section .rodata=/dev/stdout
```
lists the DTD content so looks like we're good here

File: src/registry.c (hunk 15 above)
Comment (whot) [reply src/registry.c:1141]: I'd argue against it because we rely on the DTD to validate our input files, there's no real error checking elsewhere. If this fails libxml is busted to some uncomfortable degree and skipping over this and hoping that the rest of the parser works is likely to cause us more pain than just failing.

File: src/registry.c (hunk 16 above)
Comment (whot) [reply src/registry.c:1176]: [doc says](http://xmlsoft.org/html/libxml-xmlversion.html#LIBXML_TEST_VERSION) 
> Macro to check that the libxml version in use is compatible with the version the software has been compiled against

Source code shows it does a major/minor (but not patchlevel) version comparison. Given all examples include it, I felt it was safer to do so too.

File: test/registry.c [hunk 17]
```diff
@@ -0,0 +1,843 @@
+/*
+ * Copyright © 2020 Intel Corporation
```
[the remark this answers is not in the document]
Comment (whot) [reply test/registry.c:2]: fixed, thanks

File: src/registry.c [hunk 18]
```diff
@@ -0,0 +1,1196 @@
...
+#include "config.h"
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <libxml/parser.h>
+#include <libxml/xinclude.h>
```
[the remark this answers is not in the document]
Comment (whot) [reply src/registry.c:33]: fixed, thanks

File: src/registry.c [hunk 19]
```diff
@@ -0,0 +1,1196 @@
...
+    CONTEXT_FAILED,
+};
+
+struct rxkb_context {
+    struct rxkb_object base;
+    enum context_state context_state;
+
+    bool load_extra_rules_files;
+
+    struct list models;
```
[the remark this answers is not in the document]
Comment (whot) [reply src/registry.c:77]: fixed, thanks

File: src/registry.c (hunk 14 above)
Comment (whot) [reply src/registry.c:1151]: fixed by moving the error msg out into the caller

**Comment by whot** [comment]
Merged, thanks heaps for the reviews. Anything left is easier to handle as a follow-up commit.
